### PR TITLE
[NFC][Clang] Refactor ClangDiagnosticEmitter to use more StringRef

### PR DESCRIPTION
**Pull request by jurahul**
Refactor ClangDiagnosticEmitter to use more StringRefs. Also use more range for loops.

Verified that .inc files generated by clang build are identical with and without the change.

### CODE REVIEW

**Review by kazutakahirata — COMMENTED** [review]
Is there any way you could split up into several patches?  It's hard to follow when you do multiple things in one patch.  Thanks!

**Comment by jurahul** [comment]
> Is there any way you could split up into several patches? It's hard to follow when you do multiple things in one patch. Thanks!

Let me try to do it. May be a PR for just range based for loops to start with
